File: csharp/Y2025/Day02.cs
namespace AdventOfCode.Y2025;

public sealed class Day02() : AdventDay(2025, 2)
{
    public override AdventDaySolution Solve(string input)
    { 
        var (part1, part2) = (0L, 0L);

        foreach (var (min, max) in ParseInput(input))
        {
            for (var id = min; id <= max; id++)
            {
                var idString = id.ToString();

                if (idString.Length < 2)
                    continue;

                if (!IsValid(idString, 2))
                {
                    part1 += id;
                    part2 += id;
                    continue;
                }

                if (Enumerable.Range(2, idString.Length - 1).Any(m => !IsValid(idString, m)))
                    part2 += id;
            }
        }

        return (part1, part2);
    }

    private static bool IsValid(string idString, int numParts)
    {
        if (idString.Length == 0 || idString.Length % numParts != 0)
            return true;

        var size = idString.Length / numParts;
        var part = idString[..size];
        var invalid = string.Join("", Enumerable.Repeat(part, numParts));

        return !idString.Equals(invalid);
    }

    private static IEnumerable<(long MinValue, long MaxValue)> ParseInput(string input) => input
        .Split(',')
        .Select(line => line.Split('-'))
        .Select(parts => (long.Parse(parts[0]), long.Parse(parts[1])));
}

File: fsharp/src/Y2025/Day02.fs
namespace AdventOfCode.Y2025

open AdventOfCode

type Day02() =

    let parseRange (line: string) =
        line.Split '-' 
        |> fun parts -> [parts[0] |> int64 .. parts[1] |> int64]

    let getRepeatedString part count =
        Seq.replicate count part |> String.concat ""

    let getPart (idString: string) numParts =
        idString.Length / numParts
        |> fun len -> idString[..len-1]

    let isPartRepeated (idString: string) numParts : bool =
        getPart idString numParts
        |> fun part -> getRepeatedString part numParts 
        |> idString.Equals

    let mustBeValid (idString: string) numParts =
        idString.Length = 1 || idString.Length % numParts <> 0 

    let isValid id numParts : bool =
        id
        |> string
        |> fun idString ->
            mustBeValid idString numParts ||
            isPartRepeated idString numParts |> not

    let parseIdsFromString (content: string) =
        content.Split ','
        |> Seq.collect parseRange
        |> Seq.toArray

    interface IAdventDay with
        member _.Year = 2025
        member _.Day  = 2

        member _.Solve input =
            let ids = input |> parseIdsFromString
            let part1 =
                ids
                |> Seq.filter (fun id -> isValid id 2 |> not)
                |> Seq.sum
            let part2 =
                ids
                |> Seq.filter (fun id -> (
                    let idString = id |> string
                    [2..idString.Length] |> Seq.exists(fun s -> isValid id s |> not)
                ))
                |> Seq.sum

            part1.ToString(), part2.ToString()